File: exchange/OrderBook.ts
import { Num } from "ccxt";
import { Decimal } from "decimal.js";

export default class OrderBook {
  symbol: string;
  timestamp: number;
  bids: [Decimal, Decimal][];
  asks: [Decimal, Decimal][];
  bestBid: [Decimal, Decimal];
  bestAsk: [Decimal, Decimal];
  midPrice: Decimal;
  spread: Decimal;
  spreadBps: Decimal;
  /**
   *
   */
  constructor(
    symbol: string,
    timestamp: number,
    bids: [Num, Num][],
    asks: [Num, Num][],
  ) {
    this.symbol = symbol;
    this.timestamp = timestamp;

    this.bids = bids
      .map((value) => [Number(value[0]), Number(value[1])] as [number, number])
      .sort((a, b) => b[0] - a[0])
      .map((value) => [new Decimal(value[0]), new Decimal(value[1])]);
    this.asks = asks
      .map((value) => [Number(value[0]), Number(value[1])] as [number, number])
      .sort((a, b) => a[0] - b[0])
      .map((value) => [new Decimal(value[0]), new Decimal(value[1])]);
    this.bestBid = this.bids[0];
    this.bestAsk = this.asks[0];
    this.midPrice = this.bestBid[0].add(this.bestAsk[0]).div(2);
    this.spread = this.bestAsk[0].sub(this.bestBid[0]).toDecimalPlaces(2);
    this.spreadBps = this.spread
      .div(this.midPrice)
      .mul(10_000)
      .toDecimalPlaces(2);
  }
  
}

File: exchange/order_book.py
# exchange/order_book.py

from decimal import Decimal
from typing import List, Tuple


class OrderBook:
    """
    Normalized order book representation.
    """

    def __init__(
        self,
        symbol: str,
        timestamp: int,
        bids: List[Tuple[Decimal, Decimal]],
        asks: List[Tuple[Decimal, Decimal]],
    ):
        self.symbol = symbol
        self.timestamp = timestamp

        # Convert all entries to Decimal and sort
        self.bids: List[Tuple[Decimal, Decimal]] = sorted(
            [(Decimal(str(price)), Decimal(str(qty))) for price, qty in bids],
            key=lambda x: x[0],
            reverse=True,
        )
        self.asks: List[Tuple[Decimal, Decimal]] = sorted(
            [(Decimal(str(price)), Decimal(str(qty))) for price, qty in asks],
            key=lambda x: x[0],
        )

        # Best bid / ask
        self.best_bid: Tuple[Decimal, Decimal] = (
            self.bids[0] if self.bids else (Decimal(0), Decimal(0))
        )
        self.best_ask: Tuple[Decimal, Decimal] = (
            self.asks[0] if self.asks else (Decimal(0), Decimal(0))
        )

        # Mid price and spread
        self.mid_price: Decimal = (
            (self.best_bid[0] + self.best_ask[0]) / Decimal(2)
            if self.best_bid[0] and self.best_ask[0]
            else Decimal(0)
        )
        self.spread: Decimal = (
            (self.best_ask[0] - self.best_bid[0]).quantize(Decimal("0.01"))
            if self.best_bid[0] and self.best_ask[0]
            else Decimal(0)
        )
        self.spread_bps: Decimal = (
            (self.spread / self.mid_price * Decimal(10_000)).quantize(Decimal("0.01"))
            if self.mid_price != 0
            else Decimal(0)
        )
